### PR TITLE
fix for incorrect implementation of KeyDeserialize for (T, U)

### PR DESCRIPTION
**Pull request by ewoolsey**
I think this should be sufficient to fix the problem we were seeing. There is one open question here, and that is wether or not to provide the default implementation for `KEY_LEN = 1`. I think this is unwise as it will likely result in uncaught bugs for the end user. I recommend leaving it as it is in this PR.

### CODE REVIEW

**Comment by ewoolsey** [comment]
One other option we have is make an entirely separate trait `KeyLen` which provides access to the const `KEY_LEN`. We can then require the bounds:
```rust
impl<T: KeyDeserialize + KeyLen, U: KeyDeserialize> KeyDeserialize for (T, U)
```
This has the benefit of limiting the breaking changes _only_ to projects who are very likely to be exposed to this bug.

Or project is a bit held up by this PR so it would be awesome to get this pushed through as soon as we're sure it's sound.

**Comment by maurolacy** [comment]
Thanks for the PR. I think this is good enough as it is. 

Would just like to test it a bit more before approval / merging; which can be as soon as beginning next week.

File: src/de.rs
Comment (maurolacy) [inline src/de.rs:165]: Nit: These extensions can be avoided by computing the full key length in bytes iteratively, and then doing a single vector assignment / split outside the loop.

**Review by maurolacy — APPROVED** [review]
LGTM.

File: src/de.rs
Comment (ewoolsey) [reply src/de.rs:165]: For sure you can limit the number of extensions, it does clutter up the code more though. The alternative would be using `Vec::split_off` twice (which is two allocations), calling Vec::remove twice (not very efficient), or writing a custom loop to remove the length prefix (cumbersome).

I personally would leave as is but if you have a nice way of doing it feel free!

File: src/de.rs
Comment (ewoolsey) [reply src/de.rs]: 86040d3423ea1148342fe1da8cc005da7b377e12

File: src/de.rs
Comment (maurolacy) [reply src/de.rs]: Improved naming and vars, but left the main logic intact.

**Comment by ewoolsey** [comment]
@maurolacy Thanks so much for your help! Really appreciate it. When do you think we can expect a new release on crates.io?

File: src/de.rs
Comment (webmaster128) [inline src/de.rs:13]: This is a breaking change since the trait is public

Comment (ewoolsey) [reply src/de.rs:13]: Yes it is. Unfortunately there is no solutions to this bug that does not introduce breaking changes. If you look at one of my earlier suggestions I mentioned that creating a new trait called `KeyElems`, which contains the const, and then using that as a trait bound for `KeyDeserialize for (T, U)` would make this change _less_ braking for most users. I still think that's a good solution.

Comment (webmaster128) [reply src/de.rs:13]: Thank you! We'll check next week how we can ensure the 1.x release series remains free of breaking changes. It might take a while until this can be released. So I recommend using the main branch from GitHub as a dependency to unblock you.

Comment (ewoolsey) [reply src/de.rs:13]: Will do. If you go with the new trait solution, then the changes will only be braking specifically to users who are directly exposed to the bug. Isn't it a good thing that they will now receive a compiler error, forcing them to address the issue? That is definitely the route I would take.

Comment (webmaster128) [reply src/de.rs:13]: Yeah it might be the best option. I need to look into it with a fresh head again. For now I just want to make sure this does not break happy users when released.

**Comment by webmaster128** [comment]
Thank you for your contribution, @ewoolsey & @maurolacy and sorry for my curt comment before. The change makes sense but given that this is a breaking change, I wonder what the root cause of the issue is and if there alternative ways to fix it.

Do I understand correctly that there are fundamentally two design decitions that made this change necessary?
1. When serializing tuples in keys, [they are flattened](https://github.com/CosmWasm/cw-storage-plus/blob/v1.0.1/src/keys.rs#L150-L155)
2. It is possible to use () as a key which has not even a size when serialized

Both of them are questionmarks to me. Why should (U, V, H) have 6 key elements instead of 3 when each of them is a tuple? Should this be nested the same way the Rust type is? And then large parts of the PR look like special handling for `()`. I don't understand why `()` is allowed as a key type in the fist place. I mean I can create a `Map<(), String>` but it does not make any sense. Did we go too far with the complexity here?

Anyway, given the state of things, the PR looks like great work.

**Comment by maurolacy** [comment]
> Thank you for your contribution, @ewoolsey & @maurolacy and sorry for my curt comment before. The change makes sense but given that this is a breaking change, I wonder what the root cause of the issue is and if there alternative ways to fix it.
> 
> Do I understand correctly that there are fundamentally two design decitions that made this change necessary?
> 
> 1. When serializing tuples in keys, [they are flattened](https://github.com/CosmWasm/cw-storage-plus/blob/v1.0.1/src/keys.rs#L150-L155)
> 2. It is possible to use () as a key which has not even a size when serialized

2 is not an issue here. The issue is with nested tuples used as keys. I don't understand the exact details, but basically, when nesting tuples, they are being conflated. Probably because of flattening, as you say. And a way around it, given that nested tuples are in fact being used, is this PR.

If you find an alternative (non-breaking) way to address this, we can in fact revert these changes and go for the non-breaking way.

> 
> Both of them are questionmarks to me. Why should (U, V, H) have 6 key elements instead of 3 when each of them is a tuple? Should this be nested the same way the Rust type is?

As mentioned, if you find an alternative that is breaking, that is very much welcome.

> And then large parts of the PR look like special handling for `()`. I don't understand why `()` is allowed as a key type in the > fist place. I mean I can create a `Map<(), String>` but it does not make any sense. Did we go too far with the complexity  here?

Amusingly, maps with `()` are being used; basically, to serve as a proxy or alternative for an item. This is good for generalisation. Take a look at `SnapshotItem` implementation.

> 
> Anyway, given the state of things, the PR looks like great work.

Indeed. The change may be breaking, but it is minimally so IMO. And, the entire thing with keys and keys deserialisation, including nested ones, is working, which is kind of amazing.

**Pull request merged**
